### DISCUSSION TITLE
Route-based VPN with VTI Devices - PLUTO_MARK_OUT not show the values

### DISCUSSION THREAD

**Discussion by alex-airspan** [strongswan/strongswan#1758] [Q&A]
Hi 

I run strongSwan 5.7.2 on Centos 7 with kernel 3.10.0-1160.71.1.rt56. 
I cxonfigure mark_in = mark_out = 42 according to example, but  "${PLUTO_MARK_OUT%%/*}" not show any value. 

Can you please describe how calculate the mask - 0x01000201? It dependence to  subnet's ?

Thanks,
Alex

**Comment by tobiasbrunner** [comment]
> I cxonfigure mark_in = mark_out = 42 according to example, but "${PLUTO_MARK_OUT%%/*}" not show any value.

Maybe you configured something incorrectly. Read the log and status output.



> Can you please describe how calculate the mask - 0x01000201?

What do you mean?

**Comment by alex-airspan** [comment]
Hi Tobias,

I configure below 
ip tunnel add "${VTI_IF}" local "${PLUTO_ME}" remote "${PLUTO_PEER}" mode vti key "${PLUTO_MARK_OUT%%/*}"

in log I see
07[CHD] updown: /etc/strongswan/updownScript_vti.sh: line 39: PLUTO_MARK_OUT: unbound variable
07[CHD] updown: /etc/strongswan/updownScript_vti.sh: line 39: PLUTO_MARK_OUT: unbound variable
07[CHD] updown: /etc/strongswan/updownScript_vti.sh: line 39: PLUTO_MARK_OUT: unbound variable

Charon configuration:

charon {
        install_routes=0
        load_modular = yes
        plugins {
                include strongswan.d/charon/*.conf
        }
		delete_rekeyed_delay = 60
		interfaces_use = enp5s0f0
		mark_in = mark_out = 42
		threads = 7
		cisco_flexvpn = yes
}

include strongswan.d/*.conf

[updownScript_vti.txt](https://github.com/strongswan/strongswan/files/11870802/updownScript_vti.txt)


About the second question - possible configure below. My question - how manually calculate it ?

configure mark_in = mark_out = 42 and to match the mark on ipsec0, set the value to **0x01000201** (but something like 0x00000200/0x00000f00 would also work).


Thanks,
Alex

**Reply by tobiasbrunner** [reply]
Your config is not correct. Not only is the syntax `mark_in = mark_out = 42` invalid (only a single key/value pair can be set per line), it's not even a valid option in the global strongswan.conf file. The marks are connection (or rather CHILD_SA)-specific (see [swanctl.conf](https://docs.strongswan.org/docs/5.9/swanctl/swanctlConf.html)).





> how manually calculate it ?

Marks are arbitrary 32-bit values. The example gives one in decimal notation (42) and one in hexadecimal notation (0x01000201). The latter could also be configured in decimal as 16777729, but sometimes the hexadecimal notation is easier to read, in particular if masks are involved. The configured mask (defaults to `0xffffffff`, i.e. nothing is masked) is applied first when matching the mark set on a IPsec policy/SA against the one on a packet. That's why, in the example, configuring `0x00000200/0x00000f00` works as well for packets routed via the interface that has `0x01000201` configured as key/mark. That's because after the VTI interface set the mark `0x01000201` on a packet, the mask `0x00000f00` is applied, which results in `0x00000200` and that matches the configured mark on the policies/SAs (note that this could also have been configured in decimal as `512/0x00000f00`).

**Comment by alex-airspan** [comment]
Hi Tobias, 

I add in    children  ->  net-net  -> mark_in = 42 and mark_out = 42.  Now I don't see errors and tunnel opened with mask 42

Can you please assist what preferred configuration of mark_in_sa, set_mark_in and set_mark_out  

I not fully understand from description.  

Thanks for support,
Alex

**Reply by tobiasbrunner** [reply]
> Can you please assist what preferred configuration of mark_in_sa, set_mark_in and set_mark_out

You probably don't need to set those options for the VTI use case. But what do you not understand about them?

**Comment by alex-airspan** [comment]
In document I see more different options. My question - which configuration mandatory for VTI use case.

One additional small question about PLUTO_PEER_CLIENT.  From SeGW I receive 2 different TS's. But when PLUTO_PEER_CLIENT show me only one. That it implementation of this value ? Maybe possible use with any other value?  
My current implementation - take it from swanctl --list-sas

Thanks for support,
Alex

**Reply by tobiasbrunner** [reply]
> My question - which configuration mandatory for VTI use case.

As documented, `mark_in` and `mark_out`.


> But when PLUTO_PEER_CLIENT show me only one. That it implementation of this value ? Maybe possible use with any other value?

The updown script is called for every combination of local and remote TS. So you could install the individual routes with every call and perhaps only create the VTI device if it doesn't exist yet (if you want to remove the device when the SAs are terminated you need some kind of refcounting).

Alternatively, use [vici events](https://docs.strongswan.org/docs/5.9/plugins/vici.html) where `child-updown` is called only once with all TS. There is [an example in the _route-based/net2net-xfrmi-ike_ scenario](https://github.com/strongswan/strongswan/blob/master/testing/tests/route-based/net2net-xfrmi-ike/hosts/sun/etc/updown.py) (uses XFRM interfaces but could be implemented similarly for VTIs).

**Comment by alex-airspan** [comment]
Hi Tobias.

The problem - For current I use Centos 7 with old kernel thyat not support XFRM. 

I understand from you how use updown script when received one or more subnets. Now my question - how configure with VTI interface when from Server received 0.0.0.0/0. With command - _ip route add "${PLUTO_PEER_CLIENT}" dev "${VTI_IF}"_  I got error.

You have any trick for push all traffic in tunnel ? 

Thanks,
Alex

**Reply by tobiasbrunner** [reply]
You should probably install your routes in a separate routing table to avoid conflicts with existing routes (e.g. table 220, which strongSwan uses when it installs routes) .

In general, if the peer's IP address is covered by the routes via VTI/XFRM interface, you have to avoid routing loops. You have basically two options:

* Install an explicit route to the peer that doesn't go via VTI/XFRM. For instance, install a throw route in table 220 for the peer's IP so the routes in table 220 are ignored and the main routing table is used (something like `ip route add throw $PLUTO_PEER table 220`)
* Set marks on the IKE (`charon.plugins.socket-default.fwmark = <mark>`) and ESP packets (via `set_mark_out = <mark>`), and exclude such packets from routing table 220 via a negative mark on the routing rule (`charon.plugins.kernel-netlink.fwmark = !<mark>`), where `<mark>` is an arbitrary value (but not one used already for e.g. VTIs)

The advantage of the latter is that it doesn't require a route for each peer and works even if the peer's IP changes. But it requires a relatively recent kernel (one that support `set_mark_out`).

**Comment by alex-airspan** [comment]
Hi Tobias,

For first option - I add in updown script _ip route add throw $PLUTO_PEER table 220_ and after that _ip route add "${PLUTO_PEER_CLIENT}" dev "${VTI_IF}" table 220_ . Looks that it working. That it correct ?

Thanks,
Alex

**Reply by tobiasbrunner** [reply]
Yep, that's it.

**Comment by alex-airspan** [comment]
Hi Tobias, 

The _throw_ option in route it new for me and according that I understand, when I create new table with throw - it like "replace" the main table. 
And if I want extract some networks from IPSEC, I copy subnets from main to table 220 and it looks that it working. I am correct ? 

An additional - in general - when IPSEC created , not related it I receive any subnets or receive 0.0.0.0/0 - create table 220 and add subnets to 220 - it "always" correct? I am correct ?


Below example that I copy few subnets from table main to table 220. 

[root@localhost ~]# **ip route show table 220** 
default dev ipsec0 scope link 
172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 
172.27.113.0/24 via 172.27.128.254 dev enp5s0f0 
throw 172.27.123.20 
[root@localhost ~]# 

[root@localhost ~]# **ip route show table main**
default via 172.27.128.254 dev enp5s0f0 
169.254.0.0/16 dev enp3s0 scope link metric 1002 
169.254.0.0/16 dev enp5s0f0 scope link metric 1003 
172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 
172.27.128.0/24 dev enp5s0f0 proto kernel scope link src 172.27.128.216 
172.27.129.0/24 dev ipsec0 scope link 
172.28.102.0/24 dev ipsec0 scope link 
192.168.25.0/24 dev enp3s0 proto kernel scope link src 192.168.25.1 
[root@localhost ~]# 

Thanks,
Alex

**Reply by tobiasbrunner** [reply]
> when I create new table with throw - it like "replace" the main table.

Which routing tables are used is determined by the routing policies/rules (see `ip rule`). And with the rule installed by strongSwan for table 220 the routes installed there will take precedence over the routes in the main table (so maybe you could say they "replace" them). With `throw` the table is skipped for matching destinations, so tables with lower priority (e.g. the main table) are considered and the routes there will apply.

> And if I want extract some networks from IPSEC, I copy subnets from main to table 220 and it looks that it working. I am correct ?

You don't have to copy the routes, just install a throw route for the destination so the route in the main table is used. Note that this will only  work if virtual IPs are used or for route-based setups (with a policy-based setup you also need bypass policies).

You can test if the routing works as expected with `ip route get <dest>`.

**Comment by alex-airspan** [comment]
Hi Tobias,

I think that now I understand you. Below my current updated configuration of table 220 and main table. 

I add to table 220 2 subnets throw 172.27.113.0/24 and throw 172.17.0.0/16. Both looks not forwarded over IPSEC and this is a simple way "extract" subnets from IPSEC tunnel. I am correct ?

[root@localhost ~]# **ip route show table 220**
default dev ipsec0 proto static scope link 
throw 172.17.0.0/16 
throw 172.27.113.0/24 
throw 172.27.123.20 
[root@localhost ~]# 

[root@localhost ~]# **ip route show table main**
default via 172.27.128.254 dev enp5s0f0 
172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 
172.27.128.0/24 dev enp5s0f0 proto kernel scope link src 172.27.128.216 
[root@localhost ~]#

Thanks for support,
Alex

**Reply by tobiasbrunner** [reply]
Yep, packets to these destinations are not handled by the default route in table 220 and therefore not routed via `ipsec0` device and not protected.

**Comment by alex-airspan** [comment]
Thank Tobias for perfect support.

I have additional question about re-authentication. The rekey working perfect, but when re-authentication - according to log I see that process success, but it I run UDP traffic over channel - I see that traffic dropped for some time.

Few question :
If it expected that traffic will dropped for some time? 
If need use make-before-brake configuration for resolve it ?
If it different when SeGW strongswan or SeGW Cisco IRS? If need use cisco_flexvpn = yes  for resolve it ?

Thanks,
Alex

**Reply by tobiasbrunner** [reply]
> If it expected that traffic will dropped for some time?

With break-before-make reauthentication definitely. But even with make-before-break reauthentication there could be a short interruption as the responder can install the CHILD_SA before the initiator.



> If need use make-before-brake configuration for resolve it ?

If you need reauthentication, yes. But you may also just use IKE rekeying. See [the docs](https://docs.strongswan.org/docs/5.9/config/rekeying.html) for details.

Please be aware that make-before-break reauthentication could be tricky for updown scripts (or even vici scripts, but it might be easier to keep state or do refcounting there) as the "up" even for the new IKE SA comes before the "down" event for the old SA.

>  If it different when SeGW strongswan or SeGW Cisco IRS? If need use cisco_flexvpn = yes for resolve it ?

No idea.

**Comment by alex-airspan** [comment]
Hi Tobias,

Only for clear - It possible disable re-authentication from client side and if Server configure with re-authentication - the slave will be perform it ?

Thanks,
Alex

**Reply by tobiasbrunner** [reply]
See [the docs](https://docs.strongswan.org/docs/5.9/config/rekeying.html#_reauthentication), in particular in regards to the AUTH_LIFETIME notify.
